src/mr/coordinator.go

```diff
@@ -1,40 +1,199 @@
 package mr
 
-import "log"
-import "net"
-import "os"
-import "net/rpc"
-import "net/http"
-
+import (
+	"fmt"
+
+	"log"
+	"net"
+	"net/http"
+	"net/rpc"
+	"strconv"
+	"sync"
+	"sync/atomic"
+)
 
 type Coordinator struct {
-	// Your definitions here.
+	inputfiles      []string
+	reduceTasksCnt  int
+	workerCnt       int32
+	workerStatusMap SafeMap
+	// key: string representation of task
+	// value: task status, 1-idle, 2-in progress, 3-completed
+	taskStatusMap           SafeStatusMap
+	idleMapTasksQueue       chan Task
+	idleReduceTaskQueue     chan Task
+	taskInitializationMutex sync.Mutex
+	// atomic interger
+	remainingMapTasksCnt    int32
+	remainingReduceTasksCnt int32
+}
+
+type Task struct {
+	TaskType   int // 1-map, 2-reduce
+	Inputfiles []string
+	TaskId     int
+}
+
+func (t Task) toString() string {
+	return fmt.Sprintf("%v", t)
+}
+
+type SafeMap struct {
+	m  map[string]bool
+	mu sync.Mutex
+}
+
+type SafeStatusMap struct {
```

Comment:
do we have a better way for a thread safe map? since go doesn't support generic yet, this seems really redundant